### ISSUE TITLE
Implement Subscription support on Microprofile Graphql

### ISSUE THREAD

**Issue by DorianMaliszewski** [eclipse/microprofile-graphql#174]
Hi,

Just show the documentation, and didn't see the Subscription support. I don't see it in issues too.
It can be a really good feature.🤔

What do you think about this ? 

Thanks you 😁

**Comment by andymc12** [comment]
+1 - we have discussed subscription support and I think we determined that it should go into a future release (1.1 or later).  

We've also discussed the underlying technology for subscriptions (feedback welcome).  
- WebSockets are ideal because they work two-way, but web-sockets are not part of the MicroProfile umbrella.
- Server Sent Events are part of JAX-RS (which is part of MP), so we could use SSEs, but that is only one-way (server-to-client).

**Comment by DorianMaliszewski** [comment]
Hum 🤔 According to the subscription goal, SSE is a good solution. But, implement websocket to make queries, mutations and subscription with only one connection can be good too (all pass through WS). 
If we talk only for subscription, the goal is the server need to notify changes so SSE can do the job perfectly. 👌🏻

**Comment by kaqqao** [comment]
The issue here is that while technically possible to implement subscriptions in many ways, the existing tooling currently only deals with WebSockets as Apollo's graohql-ws protocol is the only currently defined protocol for subscriptions. If we decide to eschew it, we'll have to define both a new protocol and new tooling around it... Both can be minimalistic so not necessarily a show stopper, but certainly an important consideration.

**Comment by phillip-kruger** [comment]
I think we did discuss this, but what about https://github.com/eclipse/microprofile-reactive-messaging ?

**Comment by DorianMaliszewski** [comment]
Yes it's a good alternative for sure, I just suggested to add Subscription to have Real-time when implementing GraphqlApi 😁

**Comment by phillip-kruger** [comment]
Yes I am suggesting building the subscription support on reactive messaging as that is part of the platform. So we add subscription in the spec, and when implementing it, we use reactive messaging

**Comment by andymc12** [comment]
It does seems like there are a few implementations out there that use SSE as well. See https://github.com/MikeBild/graphql-subscriptions-sse-presence-list

One advantage to SSE is that it uses HTTP and is already part of the MP umbrella (courtesy of JAX-RS) and soon to be in the MP Rest Client (see issue [11](https://github.com/eclipse/microprofile-rest-client/issues/11) and PR [266](https://github.com/eclipse/microprofile-rest-client/pull/266)).  

The approach in MP Rest Client is to use reactive streams (which is included with MP Reactive Streams Operators).  I'm not an expert on reactive messaging, but maybe it is possible that we could define subscriptions such that a user could write similar code and then choose the underlying transport mechanism (SSE, web sockets, some other transport like Kafka, etc.).

**Comment by raphael-gosteli** [comment]
I'm new to this project, but I'd like to share my point of view on this topic :)

I agree with @andymc12 to use SSE because it meets the spec's definition of using a Pub-Sub system. I think that there's no need to use WebSockets as subscriptions are read-only streams from the server to the client. 

As mentioned in the specification, there's a general problem when implementing subscriptions into a stateless service. One could just ignore this problem and delegate the "state management" to the consumer. (See http://spec.graphql.org/draft/#sec-Subscription.Supporting-Subscriptions-at-Scale)

@phillip-kruger I would say reactive messaging might not be the perfect match as you may want to consume a GraphQL service from your (JavaScript-based) frontend application without the need of 3rd-party tooling.

**Comment by DorianMaliszewski** [comment]
It's a very good idea, you're right subscription is only one way with pub-sub pattern. Reactive messaging can be extended to be use on Subscription ? I'm not an expert of GraphQL so I don't know what is the spec on it but Everytime I used it it was with Websocket.

May be in the future we will find a good solution for that. 😁
